### ISSUE TITLE
Sync / Clear bucket before uploading?

### ISSUE THREAD

**Issue by andrewmartin** [MikaAK/s3-plugin-webpack#74]
I have a really quick question; is there an option to `sync` the assets up, e.g. clean the S3 bucket before I upload? Just curious how you might handle asset revving to clear out the old assets to keep the bucket clean. Sorry if I missed something obvious, thanks again for your wonderful work on this. Using it in production with great success!

Cheers.

**Please complete these steps and check these boxes (by putting an `x` inside
the brackets) _before_ filing your issue:**

- [ x ] I have read and understood this plugin's README
- [ x ] If filing a bug report, I have included my version of node and s3-plugin-webpack
- [ x ] If filing a bug report, I have included which OS (including specific OS
  version) I am using.
- [ x ] If filing a bug report, I have included a minimal test case that reproduces
  my issue.
- [ x ] **I understand this is an open-source project staffed by someone with a job and
  that any help I receive is done on free time. I know I am not entitled to anything and will be polite and courteous.**
- [ x ] **I understand my issue may be closed if it becomes obvious I didn't
  actually perform all of these steps or the issue is not with the library itself**

Thank you for adhering to this process! This ensures that I can pay attention to issues that are relevant and answer questions faster.

# Issue Details

If filing a bug report, please include **a list of steps** that describe how to
reproduce the bug you are experiencing. Include your config being passed to the `S3Plugin`.

**Comment by MikaAK** [comment]
Honestly when releasing for prod you usually want to use a hash on the end of your filenames to cache bust. I'm fairly certain this will override the old files though

**Comment by andrewmartin** [comment]
Right, so I am actually using hashing to invalidate caches… Actually, that’s the issue. Since I’m generating unique filenames each time, it’s basically just adding more and more files to my S3 bucket. I can of course go through and manually clear them out from time to time and just rebuild, but just curious if there was an option to make it more automated.

**Comment by MikaAK** [comment]
Sadly no, though I would totally accept a pr with one since I could see this being useful! I can of course add it later on too

**Comment by andrewmartin** [comment]
Cool! I'll look into it. I know someone has a sync plugin out there.

**Comment by wootencl** [comment]
@andrewmartin Out of curiosity, did you ever come up with a solution? Dealing with this exact same issue.

**Comment by andrewmartin** [comment]
@wootencl I got sidetracked on it, ended up leaving it for now. Would love to hear if you / anyone came up with a solution...

**Comment by gertjanwytynck** [comment]
Does someone has a solution for this issue?

**Comment by MikaAK** [comment]
Nothing yet, PR's welcome!

**Comment by Birowsky** [comment]
Love message to whoever decides to roll up sleeves for this: in order to achieve zero downtime, deletion of the old/obsolete objects would need to happen after the upload of the new ones, not before.

**Comment by dejanvasic85** [comment]
Made an attempt on this here and seems to work just fine. 
https://github.com/dejanvasic85/s3-web-cleaner/blob/master/deploymentCleaner.js

> A napkin note to whoever decides to roll up sleeves for this: in order to achieve zero downtime, deletion of the old/obsolete S3 objects would need to happen after the upload of the new ones, not before.

Spot on. We are using Circle CI so this is dockerized and just runs after the deployment completes to ensure no downtime. Let me know if you guys are interested, and I can move this code inside the plugin with a config (default to false): 
```
s3Options: {
   cleanAfterDeployment: true
}
```

**Comment by MikaAK** [comment]
@dejanvasic85 A PR is welcome!

**Comment by Turbotailz** [comment]
2 years later, still no PR?

**Comment by coopersamuel** [comment]
Any news on this?

**Comment by bfagundez** [comment]
This would be nice!

**Comment by MikaAK** [comment]
> 2 years later, still no PR?

Maybe a good opportunity to do some OSS work 😉

**Comment by dejanvasic85** [comment]
Hey there, I came across needing to use this tool again. Got a branch to cater for a sync option but when I push the branch I am getting access denied @MikaAK.
